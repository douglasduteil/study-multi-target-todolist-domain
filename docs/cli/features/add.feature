Feature: dd-todo action
	In order to organize my list
	As a logged in user
	I want to be be able to create new todo lists

Background: The user has install the dd-todo
	Given I install the dd-todo

Scenario: Adding a todo list
	When I run 'dd-todo add "Hello World"'
	Then The command is successful

Scenario: Listing the todos
	When I run 'dd-todo list'
	Then The command is successful
  And I see "Hello World"
