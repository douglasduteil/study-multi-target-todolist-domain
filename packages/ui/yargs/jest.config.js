//

module.exports = {
  preset: "ts-jest/presets/js-with-ts",
  testEnvironment: "node",
  collectCoverageFrom: ["src/**/*.{ts,tsx}"],
  transformIgnorePatterns: ["node_modules/(?!ramda)"],
  globals: {
    "ts-jest": {
      tsConfig: {
        strict: false,
        allowJs: true
      }
    }
  }
};
