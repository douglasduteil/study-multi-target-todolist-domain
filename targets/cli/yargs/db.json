{
  "index": 0,
  "todos": []
}