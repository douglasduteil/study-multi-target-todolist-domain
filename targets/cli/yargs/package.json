{
  "name": "@todolist/target.yargs",
  "version": "0.0.0",
  "description": "",
  "private": true,
  "main": "index.js",
  "scripts": {
    "build": "ncc build ./src/index.ts -o dist",
    "lint": "tslint -p tsconfig.json -t stylish",
    "integration": "jest --config integration/jest.config.js",
    "test": "jest",
    "start": "ncc run ./src/index.ts",
    "watch": "yarn start --watch"
  },
  "keywords": [],
  "author": "Douglas Duteil <douglasduteil@gmail.com> (douglasduteil.github.io)",
  "license": "WTFPL",
  "dependencies": {
    "@todolist/core": "^1.0.0",
    "@todolist/databases.lowdb": "^1.0.0",
    "@todolist/ui.yargs": "^1.0.0",
    "lowdb": "^1.0.0"
  },
  "devDependencies": {
    "@types/jest": "^23.3.12",
    "@types/node": "^10.7.1",
    "@types/lowdb": "^1.0.6",
    "@zeit/ncc": "^0.8.1",
    "prettier": "^1.15.3",
    "ts-loader": "^5.3.1",
    "tslint": "^5.12.0",
    "typescript": "^3.2.2"
  }
}
