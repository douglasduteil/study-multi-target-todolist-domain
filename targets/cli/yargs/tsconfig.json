{
  "extends": "../../../tsconfig.json",
  "compilerOptions": {
    "resolveJsonModule": true
  },
  "files": ["src/index.ts"]
}
