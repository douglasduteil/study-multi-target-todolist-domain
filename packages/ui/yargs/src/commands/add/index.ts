//

import { TodoDataSource } from "@todolist/core";
import { Argv, CommandModule } from "yargs";
import { AddArguments, Aliases } from "./types";

//

export class AddCommand implements CommandModule<Argv<{}>, AddArguments> {
  public command = "add <message>";
  public describe = "Add a todo";
  public aliases: ReadonlyArray<Aliases> = ["a"];
  constructor(private todoDataSource: TodoDataSource) {}

  // ? NOTE(douglasduteil): `handler` is assigned here to preserve the `this`
  // ? `yargs` seems to call the handlers with forced `.bind(undefined, [...])`
  public handler = (args: AddArguments) => {
    this.todoDataSource.addTodo({
      completed: false,
      createdAt: new Date(0),
      id: "",
      title: args.message,
      updatedAt: new Date(0)
    });
  };
}
