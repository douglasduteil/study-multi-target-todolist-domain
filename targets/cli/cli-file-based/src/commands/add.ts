//

import { Arguments, Argv, Options } from "yargs";

//

export const aliases: ReadonlyArray<"a"> = ["a"];
export const command = "add";
export const describe = "Add todo";

// HACK(douglasduteil): CommandBuilder<any, any> ...
// I didn't find how to make tsc happy without any-s here...
// tslint:disable-next-line:typedef
export function builder<T>(yargs: Argv<T>): Argv<any> {
  return yargs.options(yargsOptions).group(Object.keys(yargsOptions), group);
}

//

export interface AddOptions {
  done: boolean;
}

export type AddArguments = Arguments<AddOptions>;

//

export const group = "Add Options";
export const yargsOptions: { [key in keyof AddOptions]: Options } = {
  done: {
    default: false,
    describe: "Show private packages that are normally hidden",
    type: "boolean"
  }
};

//
