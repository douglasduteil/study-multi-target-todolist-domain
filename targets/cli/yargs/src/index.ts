//

import { filterTodosFn } from "@todolist/core";
import { Database } from "@todolist/databases.lowdb";
import { commands, main } from "@todolist/ui.yargs";
import LowdbFileSyncStorageAdapter from "lowdb/adapters/FileSync";
import { version } from "../package.json";

//

const lowDbDataSource = new Database(
  new LowdbFileSyncStorageAdapter("db.json")
);

const { AddCommand, ListCommand } = commands;

main({ version })
  .command(new AddCommand(lowDbDataSource))
  .command(new ListCommand(filterTodosFn, console, lowDbDataSource))
  .parse(process.argv.slice(2));
