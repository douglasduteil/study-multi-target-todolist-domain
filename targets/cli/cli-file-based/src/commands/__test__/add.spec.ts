//

import yargs from "yargs";
import { builder } from "../add";

//

describe("builder", () => {
  it("should add list options to cli", () => {
    const log = jest.fn();
    builder(yargs.usage("foo add <options>")).showHelp(log as any);
    expect(log.mock.calls[0]).toMatchSnapshot();
  });
});
