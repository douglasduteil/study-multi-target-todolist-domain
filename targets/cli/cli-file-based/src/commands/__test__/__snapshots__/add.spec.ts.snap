// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builder should add list options to cli 1`] = `
Array [
  "foo add <options>

Add Options
  --done  Show private packages that are normally hidden
                                                      [boolean] [default: false]

Options:
  --help     Show help                                                 [boolean]
  --version  Show version number                                       [boolean]",
]
`;
